Fix Bech32 decoder + encoder

jonathanknowles opened the pull request:
# Issue Number

This PR fixes the following issues:

- #311 (Bech32 decoder fails for certain known-to-be-valid Bech32 strings)
- #314 (Bech32 encoder can be coerced into producing a string that can't be decoded)

# Overview

For #311:
- [x] Adds a new type `DataPart`, which wraps `[Word5]`.
- [x] Changes the types of `decode` and `encode` so that they operate on `DataPart` instead of `ByteString`.
- [x] Ensures that the round-trip relationship between `encode` and `decode` is preserved.
- [x] Provides auxilliary functions `dataPartFromBytes` and `dataPartToBytes` to convert to (and from) `ByteString` and `DataPart`:
    * `dataPartFromBytes` **pads** with trailing zeros where appropriate.
    * `dataPartToBytes` **trims** trailing zeros where appropriate.
- [x] Provides unit tests to ensure that all reference Bech32 strings mentioned in the [Bech32 standard](https://github.com/bitcoin/bips/blob/master/bip-0173.mediawiki) can be decoded successfully.

For #314:
- [x] Reworks smart constructors for `HumanReadablePart` and `DataPart` so that inputs are always converted to lower case.

parsonsmatt commented on src/Cardano/Wallet/Kernel/DB/Sqlite/Persistent/Orphans.hs:
```diff
@@ -0,0 +1,78 @@
+           Right a -> pure a
+
+instance Read Core.Address where
+    readsPrec _ = error "Needed for HttpApiData"
```
This is... dubious, but it's "fine" as long as we never actually call `Read` on this. The `Show` instance is not derived, so it's borked.

parsonsmatt commented on src/Cardano/Wallet/Kernel/DB/Sqlite/Persistent.hs:
```diff
@@ -0,0 +1,612 @@
+import qualified Pos.Chain.Txp as Txp
+import qualified Pos.Core as Core
+
+share [ mkPersist sqlSettings { mpsPrefixFields = False } , mkMigrate "migrateAll" ] [persistLowerCase|
```
This quasiquoter is what `persistent` uses to define the entity datatypes, infer migrations, and create all the query helpers and fields.

parsonsmatt commented on src/Cardano/Wallet/Kernel/DB/Sqlite/Persistent.hs:
```diff
@@ -0,0 +1,612 @@
+    txMetaTableWalletId   Core.Address   sql=meta_wallet_id
+    txMetaTableAccountIx  Word32         sql=meta_account_ix
+
+    Primary txMetaTableId txMetaTableWalletId txMetaTableAccountIx
```
These primary keys are pulled from the original schema. In reality, this should be more of a unique key, and the actual *primary* key is the `txMetaTableId` field. That's the field that the other models use to refer to the `meta`.

parsonsmatt commented on src/Cardano/Wallet/Kernel/DB/Sqlite/Persistent.hs:
```diff
@@ -0,0 +1,612 @@
+
+    Primary outputTableTxId outputTableIndex
+
+    deriving Show Generic
```
We can get automatic foreign keys if we used the `txMetaTableId` as the primary key/ID: https://stackoverflow.com/questions/30062707/foreign-key-constraints-in-yesod-persistent

Defining the `outputTableTxId` as having the type `outputTableTxId TxMetaId sql=meta_id` would automatically create the foreign key that we want.

parsonsmatt commented on src/Cardano/Wallet/Kernel/DB/Sqlite/Persistent.hs:
```diff
@@ -0,0 +1,612 @@
+-- | Simply a conveniency wrapper to avoid 'Kernel.TxMeta' to explicitly
+-- import Sqlite modules.
+newConnection :: FilePath -> IO SqlBackend
+newConnection fp = open' (mkSqliteConnectionInfo (Text.pack fp)) (\_ _ _ _ -> pure ())
```
There's a slight change to the interface of the module -- this function returns a `SqlBackend` instead of a raw SQLite `Connection`. The `SqlBackend` provides many features, like query memoization and statement finalization. Since this shouild *really* be an opaque type, and doesn't break anything else, I don't mind this interface change.

parsonsmatt commented on test/unit/Test/Spec/TxMetaStorage.hs:
```diff
@@ -175,7 +175,9 @@ spec = do
-                    (\(StorageFailure _) -> True)
+                    (\case
+                        StorageFailure _ -> True
+                        _ -> False)
```
This would fail with a `PatternMatchError` if it wasn't that constructor.

KtorZ commented:
Thanks for the PR @parsonsmatt :pray: 
I won't likely have time to look at it this week to be honest. I just skimmed through it and I like the approach already.

KtorZ commented:
Closing the PR for now. Keeping the branch for insights. We are most likely going under a rewrite of the wallet soon. This may come in handy.

jonathanknowles committed:
Rewrite Bech32 encoder and decoder in terms of base-32 (Word5) strings.

jonathanknowles committed:
Rename `mkHumanReadablePart` to `humanReadablePartFromBytes`.

jonathanknowles committed:
Remove unnecessary `bech32` prefix from functions within `Bech32.Internal`.

jonathanknowles committed:
Wrap code to meet coding standards.

KtorZ commented on lib/bech32/test/Codec/Binary/Bech32Spec.hs:
```diff
@@ -338,16 +373,28 @@ instance Arbitrary ValidBech32String where
+        len <- choose (0, 64)
+        DataPart <$> replicateM len arbitrary
+    shrink (DataPart dp)
+        | null dp   = DataPart <$> []
```
This will cause quickcheck to loop endlessly. `shrink` should **never** yield its argument!

jonathanknowles replied:
Oops, well spotted! I've (hopefully) fixed this.

KtorZ commented on lib/bech32/src/Codec/Binary/Bech32/Internal.hs:
```diff
@@ -90,7 +95,45 @@ import qualified Data.ByteString as BS
+-- This function encodes a 'ByteString' in such a way that guarantees it can be
+-- successfully decoded with the 'dataPartToBytes' function:
+--
+-- > dataPartToBytes (dataPartFromBytes b) == Just b
```
Minor remark but I believe the documentation annotation here should start with a `>>>` otherwise haddock won't display this as you expect ^.^

jonathanknowles replied:
According to the [Haddock docs](https://www.haskell.org/haddock/doc/html/ch03s08.html), both styles are possible. With a single `>` we get:

![Screenshot_2019-05-24 Codec Binary Bech32 Internal](https://user-images.githubusercontent.com/206319/58296556-c0240380-7e06-11e9-94e5-20a777ce02d5.png)

If we instead use `>>>`, then we get this:

![Screenshot_2019-05-24 Codec Binary Bech32 Internal](https://user-images.githubusercontent.com/206319/58296653-1abd5f80-7e07-11e9-8dd4-2ad909345f05.png)

I think the latter style is provided for giving examples.

KtorZ commented on lib/bech32/src/Codec/Binary/Bech32/Internal.hs:
```diff
@@ -90,7 +95,45 @@ import qualified Data.ByteString as BS
+    mempty = DataPart []
+
+instance Semigroup DataPart where
+    (DataPart d0) <> (DataPart d1) = DataPart $ d0 <> d1
```
Suggestion: use `DerivingStrategies` + `GeneralizedNewtypeDeriving`, or `DerivingVia` to get the job done by GHC :p

jonathanknowles replied:
Thanks for the suggestion. I've used `DerivingStrategies` to simplify this code.

KtorZ commented on lib/bech32/src/Codec/Binary/Bech32/Internal.hs:
```diff
@@ -90,7 +95,45 @@ import qualified Data.ByteString as BS
+    (DataPart d0) <> (DataPart d1) = DataPart $ d0 <> d1
+
+instance Show DataPart where
+    show (DataPart dp) = "DataPart " <> show ((word5ToChar Arr.!) <$> dp)
```
I think we should put that one in the coding standards:

> Never define `Show` instances by hand

`Show` instances should really just output valid Haskell data-type, and we expect `read . show = id` at all time. If we need some human-friendly representation, then use `Buildable` instance from `fmt`, that's what the instance is for. Some preludes don't even allow programmers to write their own instance of `Show` as a guard.

We only have one exception which is our `ShowFmt` class we use to force QuickCheck to display counter examples using `Buildable` instances in order to get more readable counter examples.

jonathanknowles replied:
Hi @KtorZ!

Regarding:
> Never define `Show` instances by hand

I must admit I'm not yet convinced about this (though I could be).

My original reasoning for providing a `Show` instance here is for convenience. I can type the following within `ghci`:
```hs
> decode "abcdef1qpzry9x8gf2tvdw0s3jn54khce6mua7lmqqqxw"
Right (HumanReadablePart "abcdef", DataPart "qpzry9x8gf2tvdw0s3jn54khce6mua7l")
```
It's easy to verify (visually) that the human-readable and data parts correspond to the original string (with the separator and checksum removed). The data part is rendered with the Bech32 alphabet.

The standard derived instance is much less readable:
```hs
> decode "abcdef1qpzry9x8gf2tvdw0s3jn54khce6mua7lmqqqxw"
Right (HumanReadablePart "abcdef",DataPart {getDataPart = [Word5 {getWord5 = 0},Word5 {getWord5 = 1},Word5 {getWord5 = 2},Word5 {getWord5 = 3},Word5 {getWord5 = 4},Word5 {getWord5 = 5},Word5 {getWord5 = 6},Word5 {getWord5 = 7},Word5 {getWord5 = 8},Word5 {getWord5 = 9},Word5 {getWord5 = 10},Word5 {getWord5 = 11},Word5 {getWord5 = 12},Word5 {getWord5 = 13},Word5 {getWord5 = 14},Word5 {getWord5 = 15},Word5 {getWord5 = 16},Word5 {getWord5 = 17},Word5 {getWord5 = 18},Word5 {getWord5 = 19},Word5 {getWord5 = 20},Word5 {getWord5 = 21},Word5 {getWord5 = 22},Word5 {getWord5 = 23},Word5 {getWord5 = 24},Word5 {getWord5 = 25},Word5 {getWord5 = 26},Word5 {getWord5 = 27},Word5 {getWord5 = 28},Word5 {getWord5 = 29},Word5 {getWord5 = 30},Word5 {getWord5 = 31}]})
```
> we expect `read . show == id` at all times

That does make sense. I've added a corresponding `Read` instance, defined in terms of `dataPartFromText` (and a QuickCheck property to check for round trip compatibility.)

Thoughts?

jonathanknowles replied:
@KtorZ I've actually removed the hand-written `Read` and `Show` instances in a later commit.

KtorZ reviewed (changes requested):
The change itself looks good, a few remarks about style & infinite shrinkers :sweat_smile:

jonathanknowles committed:
Add functions to convert `DataPart` to and from `Text`, using the Bech32 alphabet.

Also add an accompanying `Read` instance for `DataPart` to accompany the
existing `Show` instance.

This change also adds a QuickCheck property for the following relationship:

>>> read (show (dp :: DataPart)) == dp

jonathanknowles committed:
Encode to and decode from `Text` instead of `ByteString`.

Use `Text` as the internal representation for `HumanReadablePart` and
`DataPart`.

As a side-effect, we can discard the `Read` and `Show` instances for
`DataPart`.

jonathanknowles committed:
Fix a previously undetected bug in the specification.

The `encode` function was mistakenly expected (by the specification) to produce
an invalid Bech32 string, when given an upper-case human readable part. If this
invalid string is passed to the `decode` function, it fails (unsurprisingly) to
decode.

This change updates the specification so that the `encode` function is
now expected to produce a valid Bech32 string.

jonathanknowles committed:
Don't convert to lower-case before calling `humanReadablePartFromText`.

In the `Arbitrary` instance for `HumanReadablePart`, we don't need to
(and shouldn't) convert the generated string to lower-case before
calling `humanReadablePartFromText`.

The `humanReadablePartFromText` function itself already converts to lower case.

KtorZ reviewed (approved):
Looks good. Are you planning on making a PR upstream to fix the reference implementation ?
